prolif/protein_helper.py

```diff
@@ -0,0 +1,560 @@
+    _STANDARD_AA,
+    ATOMNAME_ALIASES,
+    MAX_AMIDE_LENGTH,
+    amber_pool,
+    charmm_pool,
+    formal_charge_special_assign_map,
+    gromos_pool,
+    opls_aa_pool,
+    resnames_map,
+)
+from prolif.molecule import Molecule
+from prolif.residue import Residue, ResidueGroup
+
+logger = logging.getLogger(__name__)
+
+
+class ProteinHelper:
+    def __init__(self, input_topology: Molecule):
+        """Initialize the ProteinHelper class.
+        This class is designed for implicit H-bond detection.
+
+        Parameters
+        ----------
+        input_topology : str or Molecule
+            The path of input protein molecule (.pdb) or a ProLIF Molecule.
+        """
+
+        # read as prolif molecule
+        if isinstance(input_topology, Molecule):
+            self.protein_mol = input_topology
+
+        elif isinstance(input_topology, str) and input_topology.endswith(".pdb"):
+            input_protein_top = Chem.MolFromPDBFile(input_topology)
+            self.protein_mol = Molecule.from_rdkit(input_protein_top)
+
+        else:
+            raise TypeError(
+                "input_topology must be a string (path to a PDB file) or "
+                "a prolif Molecule instance."
+            )
+
+    def standardize_protein(self, **kwargs) -> None:
+        """Standardize the protein molecule."""
+
+        # guess forcefield
+        conv_resnames = set(self.protein_mol.residues.name)
+        forcefield_name = self.forcefield_guesser(conv_resnames)
+
+        # standardize the protein molecule
+        new_residues = []
+        for residue in self.protein_mol.residues.values():
+            standardized_resname = self.convert_to_standard_resname(
+                resname=residue.resid.name.upper(), forcefield_name=forcefield_name
+            )
+
+            # set the standard residue name
+            for atom in residue.GetAtoms():
+                atom.GetPDBResidueInfo().SetResidueName(standardized_resname)
+            residue.resid.name = standardized_resname
+
+            # before fixing the bond orders: strict check with non-standard residues
+            assert self.check_resnames({standardized_resname}, **kwargs), (
+                f"Residue {residue.resid} is not a standard residue or "
+                "not in the templates. Please provide a custom template."
+            )
+            # soft check the heavy atoms in the residue compared to the standard one
+            if self.n_residue_heavy_atoms(
+                residue
+            ) != self.n_standard_residue_heavy_atoms(standardized_resname):
+                warnings.warn(
+                    f"Residue {residue.resid} has a different number of "
+                    "heavy atoms than the standard residue. "
+                    "This may affect H-bond detection.",
+                    stacklevel=2,
+                )
+            # fix the bond orders
+            new_residues.append(self.fix_molecule_bond_orders(residue, **kwargs))
```

Comment:
one thing to look at in the future is whether we need to run this for every single residue or if we can detect early which ones need to be fixed to save some time